polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp

```diff
@@ -75,93 +76,264 @@ void Uniformity::print(raw_ostream &os) const { os << *this; }
+  }
+
+  // If the operation only allocates memory, the value yielded is uniform.
+  if (hasSingleEffect<MemoryEffects::Allocate>(memoryEffectOp)) {
+    LLVM_DEBUG(llvm::dbgs().indent(2) << "Operation allocates a value\n");
+    return propagateAllIfChanged(results, Uniformity::getUniform());
   }
 
-  // Analyze operations with memory side effects that have uniform operands.
+  LLVM_DEBUG(llvm::dbgs().indent(2) << "Analyzing memory effects\n");
   SmallVector<MemoryEffects::EffectInstance> effects;
   memoryEffectOp.getEffects(effects);
-  for (const auto &effect : effects) {
-    Value val = effect.getValue();
+
+  for (const auto &EI : effects) {
+    Value val = EI.getValue();
     if (!val) {
       // Memory effect on anything other than a value: conservatively assume
       // the result(s) uniformity is unknown.
       LLVM_DEBUG(llvm::dbgs() << "Memory Effect on non-values found\n");
-      for (UniformityLattice *result : results)
-        propagateIfChanged(result, result->join(Uniformity::getUnknown()));
+      return propagateAllIfChanged(results, Uniformity::getUnknown());
+    }
+
+    if (!isa<MemoryEffects::Read>(EI.getEffect()))
+      continue;
+
+    // Get the reaching def. and potential reaching def. of the value 'val' and
+    // analyze them to determine its uniformity.
+    const ReachingDefinition *rdef =
+        internalSolver.lookupState<ReachingDefinition>(op);
+    if (!rdef) {
+      LLVM_DEBUG(llvm::dbgs() << "Unable to find reaching definition\n");
+      return propagateAllIfChanged(op->getResults(), Uniformity::getUnknown());
+    }
+    LLVM_DEBUG(llvm::dbgs().indent(2) << "rdef:\n" << *rdef << "\n");
+
+    using ModifiersTy = ReachingDefinition::ModifiersTy;
+    auto merge = [](std::optional<ModifiersTy> mods,
+                    std::optional<ModifiersTy> pMods) {
+      if (!mods && !pMods)
+        return mods;
+      if (mods && !pMods)
+        return mods;
+      if (pMods && !mods)
+        return pMods;
+      mods->merge(*pMods);
+      return mods;
+    };
+
+    // Merge mods and pMods together.
+    std::optional<ModifiersTy> mods =
+        merge(rdef->getModifiers(val), rdef->getPotentialModifiers(val));
+    if (!mods)
+      continue;
+
+    // If any of mods/pMods are dominated by a branch with a condition that is
+    // unknown/non-uniform the loaded value has the same uniformity.
+    SmallVector<Value> branchConditions = collectBranchConditions(*mods);
+    if (anyOfUniformityIs(branchConditions, Uniformity::Kind::Unknown)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2)
+                 << "Branch condition has unknown uniformity\n");
+      return propagateAllIfChanged(op->getResults(), Uniformity::getUnknown());
+    }
+    if (anyOfUniformityIs(branchConditions, Uniformity::Kind::NonUniform)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2) << "Branch condition non-uniform\n");
+      return propagateAllIfChanged(op->getResults(), Uniformity::getUnknown());
```

Comment:
Should this be `Uniformity::getNonUniform()` here?

Reply:
Yes, thanks for catching this!

Reply:
done